#### Read dataset and metadate accordingly

```{r}
# Govaere
cts <-utils::read.csv(file.path("..", "Data", "govaere_v3.csv"), header = TRUE, row.names = 1)
meta<-utils::read.csv(file.path("..", "Data", "govaere_meta.csv"), header = TRUE, row.names = 1)
```


```{r}
cts <-utils::read.csv(file.path("..", "Data", "FFP_cts.csv"), header = TRUE, row.names = 1)
meta<-utils::read.csv(file.path("..", "Data", "ffp_meta.csv"), header = TRUE, row.names = 1)
```

```{r}
cts <-utils::read.csv(file.path("..", "Data", "hoang_v3.csv"), header = TRUE, row.names = 1)
meta<-utils::read.csv(file.path("..", "Data", "hoang_meta.csv"), header = TRUE, row.names = 1)
```







##### Functions for data ordering
##################################################################################################################################
##################################################################################################################################

```{r}

### Write a function to cut a tree and obtain a dataframe with the sample subgroups
get_subgroups<- function(tree_hc, k){
  ####tree_hc<-hclust() object, either on columns or rows
  groups <- cutree(tree_hc, k)
  return(groups)
}

### Write a function to subset the gene expression matrix
subset_cts <- function(cts, vector){
  cts.new <- cts[vector,]
  return(cts.new)
}

### Write a function to order dataframe based on a column
order_rows<-function(meta, order_vector){### input data metafile and vector to order the rownames
  df_ordered<- meta[order(order_vector),]### Sort metadata rownames (samples) based on vector F score
  return(df_ordered)### Return meta ordered
}

### Write a function to order cts file using the meta_ordered file rownames
order_cts <- function(cts, order_vector){
  ordered_cts <- cts[order_vector]
  return(ordered_cts)
}

### Function to round values 
round_df <- function(x, digits) {
  # round all numeric variables
  # x: data frame 
  # digits: number of digits to round
  numeric_columns <- sapply(x, mode) == 'numeric'
  x[numeric_columns] <-  round(x[numeric_columns], digits)
  x
}

####Function to remove decimals from the gene names using regex
remove_decimals <-function(cts){
  genes_with_decimals<-rownames(cts)
  genes_without_decimals<- gsub("\\..*", "", genes_with_decimals)
  rownames(cts)<-genes_without_decimals
  return(cts)
}
```


###Preprocessing of the data
##################################################################################################################################
##################################################################################################################################

```{r}

## Function to filter cts file for low-gene count values (mainly zero genes)
###Filter dataset for low-counts
###>75% of the samples has values >=10
###All the gene-rows that have <=75% with values >=10 will be removed.
####I would not remove values <10 but leave them in, as they are.
####a percentage (e.g. 75%) of the samples do not have more than 10 counts.

filter_low_counts<-function(cts, filter){
  #filter = 5 or 10
  threshold <- round(length(cts) * 0.75)
  filtered_cts <- cts[rowSums(cts>=filter) >= threshold,]
  return(filtered_cts)
}


### Write a function to normalize data as follows: 
##   nCts = (raw_gene_value/total sum each sample)*average(total sum of samples)
normalize <- function(cts){
  vector <- colSums(cts)
  average_sum<-mean(colSums(cts))
  cts.sweep <- sweep(cts, MARGIN = 2, STATS =vector, '/') 
  cts.norm   <- cts.sweep * average_sum
  return(cts.norm)
}

### Write a function to normalize for healthy samples as follows:
##    2logR = log2(nCts/mean(healthy sample average per row (genes from healthy samples)))

transformation_Rlog_<-function(cts.norm){
  boolean_healthy<-meta$Fibrosis_label=="Fibrosis_0"
  healthy_samples= cts.norm[,boolean_healthy]
  healthy_samples_means <-rowMeans(healthy_samples)
  cts.r2log<-log2((cts.norm+2)/ (healthy_samples_means+2)) ### Add integer to avoid negative values
  return(cts.r2log)
}

transformation_Rlog_positive_values<-function(cts.norm){
  boolean_healthy<-meta$Fibrosis_label=="Fibrosis_0"
  healthy_samples= cts.norm[,boolean_healthy]
  healthy_samples_means <-rowMeans(healthy_samples)
  cts.r2log <- log2((cts.norm+2)/ (healthy_samples_means+2))+3### Add integer to avoid negative values
  return(cts.r2log)
}

#### Write a function to scale the data for dimensionality reduction
## scaling function as following formula:
##  scaled_value = (value - mean value)/standard deviation
scale_cts<-function(cts.r2log){
  df<-t(cts.r2log)
  mean<-mean(colSums(df))##per row genes
  sd<- sd(colSums(df))
  cts.scaled <- cts.r2log - mean/ sd
  return(t(cts.scaled))
}

```

##################################################################################################################################





###############################################################################################################################
####    PREPROCESSING RNAseq studies (low-count filter, normalization, normalization healthy, r2log transformation, scaling)
###############################################################################################################################

```{r}
###Low-count gene filtering
cts<-filter_low_counts(cts = cts, filter = 10)
dim(cts)
```

```{r}
###Order metadata by Fibrosis label
meta_ordered<-order_rows(meta = meta, order_vector = meta$Fibrosis_label)
head(meta_ordered)
```

```{r}
##FFP
#### Order gene cts file by sample names in ordered_metadata
ordered_cts<-order_cts(cts = cts, order_vector = meta_ordered$TID)##FFP
```

```{r}
##FFP
colnames(ordered_cts)==meta_ordered$TID ###FFP dataset
```


```{r}
###Govaere and Hoang
ordered_cts<-order_cts(cts = cts, order_vector = rownames(meta_ordered))###Govaere and Hoang
head(ordered_cts)
```

```{r}
#### Check order between meta ordered and cts ordered
colnames(ordered_cts)==rownames(meta_ordered)### Govaere and Hoang
```

```{r}
###Use function to normalize
normalized_cts<-normalize(cts = ordered_cts)
dim(normalized_cts)## see dimensions
head(normalized_cts)## see values
```

```{r}
###Use function to transform normalized cts to 2logR values
cts.2logr<-transformation_Rlog_(normalized_cts)
head(cts.2logr)
min(cts.2logr)
dim(cts.2logr)
```


# Use for further analyses and visualization
## cts.2logr
### Write files of each study for further visualization

```{r}
write.csv(cts.2logr, "FFP_2rlog.csv")
```


```{r}
write.csv(cts.2logr, "Govaere_2rlog_test.csv")
```


```{r}
write.csv(cts.2logr, "Hoang_2rlog.csv")
```



















